### PR TITLE
Port remaining 10 services to Trapperkeeper

### PR DESCRIPTION
**Pull request by frenchy64**
**Epic** https://github.com/threatgrid/iroh/issues/3986

It was quite difficult to split this up further and still retain some confidence of correctness, so this PR ports all remaining services to TK (10 of them by my count). Thankfully it's only +600/-500 LOC of changes, but LMK if further splitting is needed.

Please see inline comments for more guideposts.

Noteworthy:
- splits `ctia.events` into `ctia.events-service{-core}`.
- http server management moved out of `ctia.http.server` and moved to `ctia.http.server-service{-core}`.
- splits `ctia.logging` into `ctia.logging{-core}`.
- moves store management from `ctia.store` to `ctia.store-service{-core}`
- all stubs and global services removed
- attempted to update benchmarks but some of them haven't been compilable for many years (they refer to vars like `setup-ctia-atom-store!` and `setup-ctia-es-store-native!` that don't exist)

<a name="qa">[§](#qa)</a> QA
============================

No QA is needed.

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: Port remaining 10 services to Trapperkeeper
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

### CODE REVIEW

File: src/ctia/store.clj [hunk 1]
```diff
@@ -58,14 +58,6 @@
    :vulnerability []
    :weakness []})
 
-(defonce stores (atom empty-stores))
-
```
Comment (frenchy64) [inline src/ctia/store.clj:62]: Below functions moved to `ctia.store-service-core`.

File: src/ctia/task/migration/store.clj [hunk 2]
```diff
@@ -705,11 +706,8 @@ when confirm? is true, it stores this state and creates the target indices."
 (s/defn setup!
   "setup store service"
   [{{:keys [get-config]} :ConfigService :as services} :- MigrationStoreServices]
-  ;; TODO will be rolled into trapperkeeper bootstrap
```
Comment (frenchy64) [inline src/ctia/task/migration/store.clj:708]: Starting an app here turned out not to be necessary, so this impl doesn't need to change.

File: src/ctia/task/purge_es_stores.clj [hunk 3]
```diff
@@ -1,20 +1,22 @@
...
+            [ctia.stores.es.store :refer [delete-state-indexes]]
+            [puppetlabs.trapperkeeper.app :as app]))
 
 (defn setup
-  "start CTIA store service"
+  "start CTIA store service.
+  returns a tk app."
   []
   (log/info "starting CTIA Stores...")
-  (let [config (doto (p/build-init-config)
```
Comment (frenchy64) [inline src/ctia/task/purge_es_stores.clj:13]: Note: `start-ctia!*` calls `log-properties`.

File: test/ctia/entity/entities_test.clj [hunk 4]
```diff
@@ -1,11 +1,19 @@
...
             [clojure.spec.alpha :refer [gen]]
-            [clojure.spec.gen.alpha :refer [generate]]))
+            [clojure.spec.gen.alpha :refer [generate]]
+            [puppetlabs.trapperkeeper.app :as app]
+            [puppetlabs.trapperkeeper.services :refer [service-context]]))
+
+;; FIXME this seems to fail because the specs used to generate
+;; data via gen-sample-entity don't agree with the schemas
+#_
+(use-fixtures :once mth/fixture-schema-validation)
```
Comment (frenchy64) [inline test/ctia/entity/entities_test.clj:16]: This seems like a very nasty problem worth noting.

File: test/ctia/test_helpers/benchmark.clj [hunk 5]
```diff
@@ -1,30 +0,0 @@
-(ns ctia.test-helpers.benchmark
```
Comment (frenchy64) [inline test/ctia/test_helpers/benchmark.clj:1]: Unused namespace.

File: src/ctia/init.clj [hunk 6]
```diff
@@ -16,40 +16,23 @@
...
    [ctia.version :as version]
-   [ctia.flows.hooks :as h]
-   [ctia.http.server :as http-server]
-   [ctia.shutdown :as shutdown]
-   [ctia.stores.es
-    [init :as es-init]]
-   [puppetlabs.trapperkeeper.core :as tk]
-   [puppetlabs.trapperkeeper.app :as app]))
-
-(defn- get-store-types [store-kw get-in-config]
```
Comment (frenchy64) [inline src/ctia/init.clj:33]: Moved these 3 functions to `ctia.store-service-core`.

File: test/ctia/entity/web_test.clj [hunk 7]
```diff
@@ -198,8 +198,8 @@
                        :confidence "Low"
                        :source "test"
                        :tlp "green"
-                       :owner "Unknown"
-                       :groups ["Administrators"]
+                       :owner "foouser"
+                       :groups ["foogroup"]
```
Comment (frenchy64) [inline test/ctia/entity/web_test.clj:202]: I have no idea what triggered this test difference. Does anyone know if this change makes sense? Is it a regression?

Looking at the surroundings, it seems like "foouser" and "foogroup" are better expected values here, but I'm not familiar with the code.

```clojure
     (helpers/set-capabilities! "foouser" ["foogroup"] "user" all-capabilities)
     (whoami-helpers/set-whoami-response "45c1f5e3f05d0" "foouser" "foogroup" "user")
```

File: src/ctia/events_service_core.clj [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            [ctim.schemas.common :as c]
+            [clojure.core.async :as a :refer [go-loop alt! chan tap]]
+            [schema.core :as s]
+            [schema-tools.core :as st])
+  (:import [clojure.core.async.impl.protocols Channel]
+           [java.util Map]))
+
+(def shutdown-max-wait-ms (* 1000 60 60))
+
+(defn stop
```
Comment (frenchy64) [inline src/ctia/events_service_core.clj:14]: Based on `ctia.events/shutdown!`.

File: src/ctia/events_service_core.clj [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+   to flush."
+  [{:keys [central-channel] :as context}]
+  (some->> central-channel (la/shutdown-channel shutdown-max-wait-ms))
+  (dissoc context :central-channel))
+
+(defn init [context]
+  (assoc context
+         :central-channel (la/new-channel)))
+
+(s/defn send-event
```
Comment (frenchy64) [inline src/ctia/events_service_core.clj:25]: Based on `ctia.events/send-event`.

File: src/ctia/events_service_core.clj [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  ([{:keys [central-channel :as context]} event :- es/Event]
+   (send-event context central-channel event))
+  ([_context_
+    {ch :chan :as echan} :- la/ChannelData
+    {:keys [owner timestamp http-params] :as event} :- es/Event]
+   (assert owner "Events cannot be registered without user info")
+   (let [event (if timestamp event (assoc event :timestamp (time/now)))]
+     (a/>!! ch event))))
+
+(s/defn register-listener :- Channel
```
Comment (frenchy64) [inline src/ctia/events_service_core.clj:36]: Based on `ctia.events/register-listener`.

File: src/ctia/logging.clj [hunk 11]
```diff
@@ -1,11 +1,11 @@
...
-            [clojure.tools.logging :as log]
-            [ctia.events :as e]
-            [ctia.shutdown :as shutdown]))
+  (:require [ctia.logging-core :as core]
+            [puppetlabs.trapperkeeper.core :as tk]))
 
-(defn init!
-  "Sets up logging of all events"
-  []
-  (let [control (e/register-listener #(log/info "event:" %) :blocking)]
```
Comment (frenchy64) [inline src/ctia/logging.clj:10]: Moved to `ctia.logging-core/start`.

File: src/ctia/store_service_core.clj [hunk 12]
```diff
@@ -0,0 +1,47 @@
+(ns ctia.store-service-core
+  (:require [clojure.string :as str]
+            [ctia.properties :as p]
+            [ctia.store :refer [empty-stores]]
+            [ctia.stores.es.init :as es-init]))
+
```
Comment (frenchy64) [inline src/ctia/store_service_core.clj:6]: These functions are from `ctia.init` and `ctia.store`, I left comments there as pointers.

File: src/ctia/task/check_es_stores.clj [hunk 13]
```diff
@@ -40,9 +40,9 @@
   returns trapperkeeper app"
   []
   (log/info "starting CTIA Stores...")
-  (let [config (doto (p/build-init-config)
-                 log-properties)]
-    (init-store-service! (partial get-in config))))
+  (let [config (p/build-init-config)]
```
Comment (frenchy64) [inline src/ctia/task/check_es_stores.clj:43]: `start-ctia!*` performs a `log-properties` call.

File: src/ctia/task/codegen.clj [hunk 14]
```diff
@@ -2,9 +2,8 @@
   (:require [clojure.java
              [io :as io]
              [shell :as shell]]
-            [ctia
-             [init :refer [start-ctia!]]
-             [properties :as p]]))
+            [ctia.init :refer [start-ctia!]]
+            [puppetlabs.trapperkeeper.app :as app]))
```
Comment (frenchy64) [inline src/ctia/task/codegen.clj:6]: Eep, missed aliasing app here!

File: src/ctia/task/rollover.clj [hunk 15]
```diff
@@ -41,12 +42,19 @@
           stores))
 
 (defn -main [& _args]
-  (let [config (doto (p/build-init-config)
```
Comment (frenchy64) [inline src/ctia/task/rollover.clj:44]: `log-properties` now performed in `start-ctia!*`.

File: src/ctia/task/update_mapping.clj [hunk 16]
```diff
@@ -47,10 +49,9 @@
         (println summary)
         (System/exit 0))
       (pp/pprint options)
-      (let [config (doto (p/build-init-config)
-                     init/log-properties)
```
Comment (frenchy64) [inline src/ctia/task/update_mapping.clj:51]: `log-properties` now performed in `start-ctia!*`.

File: src/ctia/init.clj [hunk 17]
```diff
@@ -93,55 +84,11 @@
 (defn start-ctia!
   "Does the heavy lifting for ctia.main (ie entry point that isn't a class).
   Returns the Trapperkeeper app."
- ([] (start-ctia! (p/build-init-config)))
- ([config]
+  []
   (log/info "starting CTIA version: "
             (version/current-version))
 
-  ;; shutdown hook
```
Comment (frenchy64) [inline src/ctia/init.clj:101]: All of this is moved into the respective services' lifecycle functions

File: test/ctia/test_helpers/core.clj [hunk 18]
```diff
@@ -35,51 +32,12 @@
...
-                              {:app app
-                               :service svc-kw})))]
-      service-map)))
-
-(def fixture-property
-  (mth/build-fixture-property-fn PropertiesSchema))
+  {:pre [(keyword? svc-kw)]}
+  (let [graph (app/service-graph app)
+        m (svc-kw graph)]
+    (assert (map? m) (str "No service " svc-kw ", found " (keys graph)))
```
Comment (frenchy64) [inline test/ctia/test_helpers/core.clj:38]: Moved `:post` check here.

File: benchmarks/ctia/entity/actor_bench.clj [hunk 19]
```diff
@@ -13,7 +13,7 @@
 
 (defgoal create-actor "Create Actor"
   :setup (fn [] [(setup-ctia-es-store!)])
-  :cleanup (fn [_] cleanup-ctia!))
+  :cleanup (fn [{:keys [app]}] (cleanup-ctia! app)))
```
Comment (frenchy64) [inline benchmarks/ctia/entity/actor_bench.clj:16]: AFAICT you have to actually execute the cleanup here https://github.com/davidsantiago/perforate/blob/61bf670d806e6696785324d89e0fcac5ffad205d/src/perforate/core.clj#L22-L23

File: test/ctia/test_helpers/benchmark.clj [hunk 20]
```diff
@@ -1,30 +1,34 @@
...
 
 (defn setup-ctia-es-store! []
   (setup-ctia! esh/fixture-properties:es-store))
 
 (def delete-store-indexes esh/delete-store-indexes)
 
-(defn cleanup-ctia! []
+(defn cleanup-ctia! [app]
   (esh/delete-store-indexes false)
-  #(shutdown/shutdown-ctia!))
```
Comment (frenchy64) [inline test/ctia/test_helpers/benchmark.clj:30]: The cleanup function doesn't return a function, this seems like an typo https://github.com/davidsantiago/perforate/blob/61bf670d806e6696785324d89e0fcac5ffad205d/src/perforate/core.clj#L22-L23

File: src/ctia/lib/metrics/console.clj [hunk 21]
```diff
@@ -1,9 +1,18 @@
...
     (when enabled
       (console/start interval))))
+
+(defprotocol ConsoleMetricsService)
+
+(tk/defservice console-metrics-service
+  ConsoleMetricsService
+  [[:ConfigService get-in-config]]
+  (start [this context]
+         (init! get-in-config)
```
Comment (ereteog) [inline src/ctia/lib/metrics/console.clj:17]: reading the `init!` function, I am wondering if it should not be renamed `start!`

File: src/ctia/lib/metrics/riemann.clj [hunk 22]
```diff
@@ -10,3 +10,12 @@
...
       (riemann/start (select-keys config
                                   [:host :port :interval-in-ms])))))
+
+(defprotocol RiemannMetricsService)
+
+(tk/defservice riemann-metrics-service
+  RiemannMetricsService
+  [[:ConfigService get-in-config]]
+  (start [this context]
+         (init! get-in-config)
```
Comment (ereteog) [inline src/ctia/lib/metrics/riemann.clj:20]: rename init! into start!?

File: src/ctia/lib/metrics/jmx.clj [hunk 23]
```diff
@@ -1,7 +1,16 @@
...
   (when (get-in-config [:ctia :metrics :jmx :enabled])
     (jmx/start)))
+
+(defprotocol JMXMetricsService)
+
+(tk/defservice jmx-metrics-service
+  JMXMetricsService
+  [[:ConfigService get-in-config]]
+  (start [this context]
+         (init! get-in-config)
```
Comment (ereteog) [inline src/ctia/lib/metrics/jmx.clj:15]: rename  `init!` into `start!`?

File: src/ctia/task/migration/store.clj [hunk 24]
```diff
@@ -55,7 +55,8 @@
 
 (def timeout (* 5 60000))
 (def es-max-retry 3)
-(def migration-es-conn (atom nil))
+;;FIXME refactor to local argument
```
Comment (ereteog) [inline src/ctia/task/migration/store.clj:58]: I created this issue for me and added it to the ES7 migration milestone.
https://github.com/threatgrid/iroh/issues/4230

File: benchmarks/ctia/tasks/migrate_es_stores_bench.clj [hunk 25]
```diff
@@ -62,18 +62,17 @@
...
+             (println "posting fixtures...")
+             (post-fixtures)
+             (prn "fixtures loaded!")
+             [app]))
+  :cleanup (fn [app]
              (println "cleanup and exit CTIA...")
-             (cleanup-ctia!)))
+             (cleanup-ctia! app)))
 
 (defn run-migrate-es-stores []
```
Comment (ereteog) [inline benchmarks/ctia/tasks/migrate_es_stores_bench.clj:77]: I will update this benchmark to make it comply to the migration changes: https://github.com/threatgrid/iroh/issues/4231

**Review by vemv — APPROVED** [review]
🎊

**Review by ereteog — APPROVED** [review]
very nice!

File: src/ctia/lib/metrics/console.clj (hunk 21 above)
Comment (frenchy64) [reply src/ctia/lib/metrics/console.clj:17]: Yes will do.

File: src/ctia/task/migration/store.clj (hunk 24 above)
Comment (frenchy64) [reply src/ctia/task/migration/store.clj:58]: Thanks!

File: benchmarks/ctia/tasks/migrate_es_stores_bench.clj (hunk 25 above)
Comment (frenchy64) [reply benchmarks/ctia/tasks/migrate_es_stores_bench.clj:77]: Thanks!

**Pull request merged**